cli/azd/cmd/container.go

```diff
@@ -307,6 +311,23 @@ func registerCommonDependencies(container *ioc.NestedContainer) {
 	container.RegisterSingleton(swa.NewSwaCli)
 	container.RegisterSingleton(terraform.NewTerraformCli)
 
+	// Provisioning
+	container.RegisterTransient(provisioning.NewManager)
```

Comment:
Do the provisioning manager and providers need to be transient or can they be singletons?  It's hard for me to understand what drives the lifetime decisions here.  My understanding is that the transient nature means that if you resolve a type from the container multiple times, you'll get back a separate instance each time, vs singleton where the instance is cached?

Is this something to do with the fact that the manager has the provider as state and the provider type could vary?

Reply:
In this case the provider would hold some transient state such as the project path and infra options.  Singleton would likely be OK now but as we expand out support for each service to have its own infra then at that point we would need a new instance of the provider.  

Another alternative would be to pass the transient state into the method params, but I feel that it complicates the API surface more than is needed.

Reply:
> In this case the provider would hold some transient state such as the project path and infra options.

Ahh - I see now - the `Provider` interfaced gained a new `Initialize` method that sets some state after construction, and so these are better represented as transient.  Makes senses.

Reply:
Curious, how do we avoid problems such as ones highlighted in https://github.com/Azure/azure-dev-pr/issues/1509 with this model?

Personally, I would love to see us avoid injecting largely stateful components that have a wide reach. Calling `manager.Initialize()` vs. not calling `manager.Initialize()` changes the behavior of the entire program entirely and is non-trivial to spot in reviews. Registering as transient vs. singleton didn't actually change this behavior, it just changed the metadata slightly. The component is still `new`'d once in cobra_builder, and stored throughout the lifetime.

I'd prefer if we just "new'd" it once explicitly for components that cannot be tied to the command lifetime. We instead register `provisioningManagerOptions` as the options for the dependencies, so that we avoid leakage of the dependencies in the command layer.

Reply:
This matches the model that we have in place for the `Project` and `Service` managers which enables us to have a single place where all required validations are performed.

Today there wouldn't be any change between `Singleton` vs `Transient` registration but when we enable per service IaC provisioning then transient would allow each service provisioning step to get a new Provision manager keeping the transient state isolated but still reference the same singleton dependences.

DI is the method that we adopted as a team and registering the manager and the providers via DI makes the model so much cleaner and allows us to remove dependencies through the command just to only use when instantiating the provisioning manager.  It also enables IaC providers to specify only the dependencies that they need vs having a single `NewProviderFn` contract that we had before.

Reply:
> I'd rather not dive into what-if scenarios here about the future of the 3rd party plugin model.

I only brought it up because that's what we did when we make bicep and terraform plugins. We created a plugin model without the need for a plugin model.

I'd also like to stake my claim and move on from this.

My claim is that code that looks like:

```golang
options := FooOptions{}
foo := NewFoo()

foo.State() // error, you didn't call Initialize()

foo.Initialize(options) // OK, now you can use foo

foo.State() // this is now OK
```

Is hard to maintain. This is what the current implementation looks like once you look closely enough. DI doesn't play a role here. But that's what we're doing by stuffing stuff like `provisioning.Manager` into DI.

Instead, code that looks like:

```golang
options := FooOptions{}
foo := NewFoo(options)
foo.State()
```

is harder to make mistakes.

Thus, we should inject `FooOptions` and not `Foo` directly into the container registration, and have `NewFoo` happen once instead of `NewFoo` then `Initialize`